### PR TITLE
Fix how the fast parser interprets string literals in Python 2

### LINKED ISSUE #1540: mypy --fast-parser fails with .encode and .decode
```
mypy --py2 --fast-parser -c 's = u"abc"; s = s.encode("utf-8").decode("utf-8");'
```

gives the error
```
<string>:1: error: Incompatible types in assignment
(expression has type "unicode", variable has type "str")
```

I get this error only when both `--py2` and `--fast-parser` are specified (i.e no errors are thrown in python 3 mode (both with and without --fast-parser) and no errors are thrown in python 2 mode without fast-parser).

### PR DESCRIPTION
**Pull request by ddfisher**
Fix #1540.

### CODE REVIEW

File: mypy/fastparse.py [hunk 1]
```diff
@@ -100,11 +101,15 @@ def find(f: Callable[[T], bool], seq: Sequence[T]) -> T:
 
 
 class ASTConverter(ast35.NodeTransformer):
-    def __init__(self, pyversion: Tuple[int, int], custom_typing_module: str = None) -> None:
+    def __init__(self,
+                 pyversion: Tuple[int, int],
+                 is_stub,
```
Comment (gvanrossum) [inline mypy/fastparse.py]: Need type annotation here.

**Comment by ddfisher** [comment]
Oops!  Added.

File: mypy/fastparse.py [hunk 2]
```diff
@@ -647,13 +652,25 @@ def visit_Num(self, n: ast35.Num) -> Node:
...
+            # because mypy considers str literals to be compatible with
+            # unicode.
+            return StrExpr(n.s)
+        else:
+            return UnicodeExpr(n.s)
 
     # Bytes(bytes s)
     @with_line
     def visit_Bytes(self, n: ast35.Bytes) -> Node:
         # TODO: this is kind of hacky
```
Comment (gvanrossum) [inline mypy/fastparse.py]: Do you still consider this hacky? I propose to remove the comments from the bodies of visit_Str() and visit_Bytes() and instead adding a block comment before the former explaining what's going on with both. Also remove the confusing remark "need to specify raw, unicode, etc?" from the former.

I also propose to update mypy/nodes.py with a block comment before class StrExpr explaining what each is used for, e.g. like this:

- Python 2
  b'x', 'x' -> StrExpr
  u'x' -> UnicodeExpr

- Python 3
  b'x' -> BytesExpr
  'x', u'x' -> StrExpr

(assuming I got it right).

It wouldn't hurt repeating all that information in typed_ast/conversions.py just before visit_Str(), plus the info how ast27 maps literals to nodes:

  b'x', 'x' -> Str(s=b'x')
  u'x' -> Str(s=u'x')

I know the u prefix is redundant in Python 3, but it's still a helpful reminder

Comment (ddfisher) [inline mypy/fastparse.py]: The "this is hacky" comment is in kind of a weird place now.  It refers to the slicing that happens in the following lines.

That comment cleanup sounds nice.  I'll add that to the PR.

File: mypy/fastparse.py [hunk 3]
```diff
@@ -647,13 +652,25 @@ def visit_Num(self, n: ast35.Num) -> Node:
     # Str(string s) -- need to specify raw, unicode, etc?
     @with_line
     def visit_Str(self, n: ast35.Str) -> Node:
-        return StrExpr(n.s)
+        if self.pyversion[0] >= 3 or self.is_stub:
+            # Hack: assume all string literals in Python 2 stubs are normal
```
Comment (gvanrossum) [inline mypy/fastparse.py]: This comment confuses me by speaking of "all string literals". In common parlance, b'x' is still considered a "string literal", even though in the narrower sense it is a "bytes literal". Having digested the code that funnels to this point I think by the time we get here, these are the possibilities:

- Python 2 module (parsed with ast27, then converted with py2to3):
  Str() is only used for u'x' literals
  Bytes() is used for 'x' and b'x' literals (which mean the same in Python 2)

- Python 3 module or stub, or stub for Python 2 (all parsed with ast35):
  Str() is used for 'x' and u'x' literals (which mean the same in Python 3)
  Bytes() is used for b'x' literals

Mypy uses a different mapping (which I wrote down in my earlier comment about the block below). So indeed the mapping here is correct. I would summarize it as follows:

- Python 2
  u'' -> Str() -> UnicodeExpr()
  b'', '' -> Bytes() -> StrExpr()

- Python 3
  b'' -> Bytes() -> BytesExpr()
  u'', '' -> Str() -> StrExpr()

**Comment by gvanrossum** [comment]
I think the code is correct. Please update the comments to make the various mappings clearer.

**Pull request merged**